docs/adr/0006_component_constructors.md

```diff
@@ -0,0 +1,334 @@
+// with the relation `local`, the media type derived from blob.MediaTypeAware, and the resource version defaulted
+// to the component version.
+type SourceInputMethodResult struct {
+    ProcessedSource *descriptor.Source
+    ProcessedBlobData blob.ReadOnlyBlob
+}
+
+
+// SourceInputMethod is the interface for processing a source with an input method declared as per
+// [spec.Source.Input]. Note that spec.Source's who have their access predefined, are never processed
+// with a ResourceInputMethod, but are directly added to the component version repository.
+// any spec.Source passed MUST have its [spec.Source.Input] field set.
+// If the input method does not support the given input specification it MAY reject the request
+//
+// The method will get called with the raw specification specified in the constructor and is expected
+// to return a ResourceMethodResult or an error.
+//
+// A method can be supplied with credentials from any credentials system by requesting a consumer identity with
+// SourceInputMethod.GetCredentialConsumerIdentity. The resulting identity MAY be used to uniquely identify the consuming
+// method and to request credentials from the credentials system. The credentials system is not part of this interface and
+// is expected to be supplied by the caller of the input method.
+//
+// The resolved credentials MAY be passed to the input method via the credentials map, but a method MAY
+// work without credentials as well.
+type SourceInputMethod interface {
+    GetCredentialConsumerIdentity(ctx context.Context, source *constructor.Source) (identity runtime.Identity, err error)
+    ProcessSource(ctx context.Context, source *constructor.Source, credentials map[string]string) (result *SourceInputMethodResult, err error)
+}
+```
+
+The contract explicitly opens the possibility to integrate our existing systems such as our credential graph.
+However, because methods must be available at runtime to the constructor library, we also support a set of provider
+interfaces that can be integrated with our plugin system.
+
+```go
+type ResourceInputMethodProvider interface {
+    // GetResourceInputMethod returns the input method for the given resource constructor specification.
+    GetResourceInputMethod(ctx context.Context, resource *constructor.Resource) (input.ResourceInputMethod, error)
+}
+
+type SourceInputMethodProvider interface {
+    // GetSourceInputMethod returns the input method for the given source constructor specification.
+    GetSourceInputMethod(ctx context.Context, resource *constructor.Resource) (input.ResourceInputMethod, error)
+}
+```
+
+To allow for actual construction of component versions, the constructor library provides a central entry function:
+
+```go
+// TargetRepository defines the interface for a target repository that can store component versions created 
+// by the constructor but also act as a target for the constructor to add resources to.
```

Comment:
I'm not sure I understand this sentence. So does this literally mean that the TargetRepository can add components + resources? If yes, then let's simplify this sentence a bit?